nano/node/node.cpp

```diff
@@ -1120,72 +1117,152 @@ nano::signature_checker::~signature_checker ()
 	(void)code;
-	release_assert (std::all_of (check_a.verifications, check_a.verifications + check_a.size, [](int verification) { return verification == 0 || verification == 1; }));
-	check_a.promise->set_value ();
+
+	return std::all_of (check_a.verifications + start_index, check_a.verifications + start_index + size, [](int verification) { return verification == 0 || verification == 1; });
 }
 
-void nano::signature_checker::run ()
+/* This operates on a number of signatures of size (num_batches * batch_size) from the beginning of the check_a pointers.
+ * Caller should check the value of the promise which indicateswhen the work has been completed.
+ */
+void nano::signature_checker::verify_async (nano::signature_check_set & check_a, size_t num_batches, std::promise<void> & promise)
 {
-	nano::thread_role::set (nano::thread_role::name::signature_checking);
-	std::unique_lock<std::mutex> lock (mutex);
-	started = true;
+	auto task = std::make_shared<Task> (check_a, num_batches);
+	++tasks_remaining;
 
-	lock.unlock ();
-	condition.notify_all ();
-	lock.lock ();
+	for (size_t batch = 0; batch < num_batches; ++batch)
+	{
+		auto size = batch_size;
+		auto start_index = batch * batch_size;
 
-	while (!stopped)
+		boost::asio::post (thread_pool, [this, task, size, start_index, &promise] {
+			auto result = this->verify_batch (task->check, start_index, size);
+			release_assert (result);
+
+			if (--task->pending == 0)
+			{
+				--tasks_remaining;
+				promise.set_value ();
+			}
+		});
+	}
+}
+
+// Set the names of all the threads in the thread pool for easier identification
+void nano::signature_checker::set_thread_names (unsigned num_threads)
+{
+	auto ready = false;
+	auto pending = num_threads;
+	std::condition_variable cv;
+	std::mutex mutex_l;
+	std::vector<std::promise<void>> promises (num_threads);
+	std::vector<std::future<void>> futures;
+	futures.reserve (num_threads);
+	std::transform (promises.begin (), promises.end (), std::back_inserter (futures), [](auto & promise) {
+		return promise.get_future ();
+	});
+
+	for (auto i = 0u; i < num_threads; ++i)
 	{
-		if (!checks.empty ())
-		{
-			auto check (checks.front ());
-			checks.pop_front ();
-			lock.unlock ();
-			verify (check);
-			condition.notify_all ();
-			lock.lock ();
-		}
-		else
-		{
-			condition.wait (lock);
-		}
+		boost::asio::post (thread_pool, [&cv, &ready, &pending, &mutex_l, &promise = promises[i] ]() {
+			std::unique_lock<std::mutex> lk (mutex_l);
+			nano::thread_role::set (nano::thread_role::name::signature_checking);
+			if (--pending == 0)
+			{
+				// All threads have been reached
+				ready = true;
+				lk.unlock ();
+				cv.notify_all ();
+			}
+			else
+			{
+				// We need to wait until the other threads are finished
+				cv.wait (lk, [&ready]() { return ready; });
+			}
+			promise.set_value ();
+		});
+	}
+
+	// Wait until all threads have finished
+	for (auto & future : futures)
+	{
+		future.wait ();
 	}
+	release_assert (pending == 0);
```

Comment:
Maybe a normal debug assert is sufficient here as it's not critical to the node's operation that threads have names.

Reply:
done